airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/source.py

```diff
@@ -0,0 +1,397 @@
+    # But in case we want to retrieve 2 levels of parent ids (For example we run Creatives stream - it needs the ad_account_ids from AdAccount
+    # and AdAccount need organization_ids from Organizations and first we must get all organization_ids
+    # and for each of them get the ad_account_ids) so switch goes to if claus to get all the nested ids.
+    # Let me visualize this for you:
+    #
+    #           organization_id_1                      organization_id_2
+    #                 / \                                    / \
+    #                /   \                                  /   \
+    # ad_account_id_1     ad_account_id_2    ad_account_id_3     ad_account_id_4
+    #
+    # So for the AdAccount slices will be [{'organization_id': organization_id_1}, {'organization_id': organization_id_2}]
+    # And for the Creatives (Media, Ad, AdSquad, etc...) the slices will be
+    # [{'ad_account_id': ad_account_id_1}, {'ad_account_id': ad_account_id_2},
+    #  {'ad_account_id': ad_account_id_3},{'ad_account_id': ad_account_id_4}]
+    #
+    # After getting all the account_ids, they go as slices to Creatives (Media, Ad, AdSquad, etc...)
+    # and are used in the path function as a path variables according to the API docs
+
+    depend_on_stream = depends_on_stream(**depends_on_stream_config)
+    depend_on_stream_slices = depend_on_stream.stream_slices(sync_mode=SyncMode.full_refresh)
+    depend_on_stream_ids = []
+
+    if depend_on_stream_slices != default_stream_slices_return_value:
+        for depend_on_stream_slice in depend_on_stream_slices:
+            records = depend_on_stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=depend_on_stream_slice)
+            depend_on_stream_ids += [{slice_key_name: record["id"]} for record in records]
+    else:
+        depend_on_stream_ids = [{slice_key_name: record["id"]} for record in depend_on_stream.read_records(sync_mode=SyncMode.full_refresh)]
+
+    if not depend_on_stream_ids:
+        return []
+
+    auxiliary_id_map[depends_on_stream.__name__] = depend_on_stream_ids
+    return depend_on_stream_ids
+
+
+class SnapchatMarketingStream(HttpStream, ABC):
+    url_base = "https://adsapi.snapchat.com/v1/"
+    primary_key = "id"
+
+    @property
+    def response_root_name(self):
+        """ Using the class name in lower to set the root node for response parsing """
+        return self.name
+
+    @property
+    def response_item_name(self):
+        """
+        Used as the second level node for response parsing. Usually it is the response_root_name without last char.
+        For example: response_root_name = organizations, so response_item_name = organization
+        The [:-1] removes the last char from response_root_name: organizations -> organization
+         {
+            "organizations": [
+                {
+                    "organization": {....}
+                }
+            ]
+        }
+        """
+        return self.response_root_name[:-1]
```

Comment:
Just saw you handle this with a property override in Media. Maybe it would be more clear to explicitly define the response_item_name per stream but I think as long as we're catching the case where we've got it wrong for any reason then it's safe.

Reply:
First I wanted to update the `parse_response` method to write log WARN message in case the response_item_name is incorrect. But after testing with incorrect response_item_name in any case we got 
```
pydantic.error_wrappers.ValidationError: 1 validation error for AirbyteRecordMessage
none is not an allowed value (type=type_error.none.not_allowed)
```

So I think the best way will be to raise an exception in case the response_item_name is absent and this logic will be untouched

Reply:
Maybe you were trying to pass None in the log message or something?

In any case, I agree with `So I think the best way will be to raise an exception in case the response_item_name is absent`. This would mean if it does happen we'll fail out and stop the state from getting updated so data won't be missed for the user.